File: producer-consumer-finite-buffer/pcfb.py
# This implementation uses Python's syncronized queue.

from threading import Thread
import queue
import time
import random

BUF_LEN = 5
NUM_PRODUCERS = 50
NUM_CONSUMERS = 2
NUM_TRIALS = 100

startTimes = [0]*NUM_PRODUCERS
endTimes = [0]*NUM_PRODUCERS

def produceEvent(event):
    time.sleep(0.001*random.randint(0,100)) # Simulate producing event.

def consumeEvent(event):
    time.sleep(0.001*random.randint(0,100)) # Simulate consuming event.

def producer(event, buffer):
    global startTimes
    if event is not None:
        #produceEvent(event)
        startTimes[event] = time.time()
    buffer.put(event)
    # if event is not None:
    #     print("Event", event, "placed on buffer.")

def consumer(buffer):
    global endTimes
    while True:
        event = buffer.get()
        if event is None:
            break
        # print("Event", event, "removed from buffer.")
        endTimes[event] = time.time()
        # consumeEvent(event)
        
def main():
    averageDurations = [0]*NUM_TRIALS
    averageWaitTimes = [0]*NUM_TRIALS
    for j in range(NUM_TRIALS):
        buffer = queue.Queue(BUF_LEN) # Finite-sized buffer.

        start = time.time()
        consumerThreads = []
        for i in range(NUM_CONSUMERS):
            consumerThread = Thread(target=consumer, args=(buffer,))
            consumerThreads.append(consumerThread)
            consumerThread.start()

        producerThreads = []
        for i in range(NUM_PRODUCERS):
            producerThread = Thread(target=producer, args=(i,buffer))
            producerThreads.append(producerThread)
            producerThread.start()

        for thread in producerThreads:
            thread.join()

        for i in range(NUM_CONSUMERS):
            Thread(target=producer, args=(None,buffer)).start()

        for thread in consumerThreads:
            thread.join()
        end = time.time()

        global startTimes
        global endTimes
        waitTimes = [endTimes[i] - startTimes[i] for i in range(NUM_PRODUCERS)]
        averageWaitTimes[j] = sum(waitTimes)/len(waitTimes)
        averageDurations[j] = end-start

    print("Average wait time:", sum(averageWaitTimes)/len(averageWaitTimes), "s")
    print("Average duration:", sum(averageDurations)/len(averageDurations), "s")


if __name__ == "__main__":
    main()

File: producer-consumer-finite-buffer/pcfb.go
package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

const bufferLength = 5 // Size of the finite buffer.
const numProducers = 50
const numConsumers = 2
const numTrials = 100

var startTimes [numProducers]time.Time
var endTimes [numProducers]time.Time

func produceEvent(event int) {
	time.Sleep(time.Duration(rand.Intn(100)) * time.Millisecond) // Simulates time to produce the event.
}

func consumeEvent(event int) {
	time.Sleep(time.Duration(rand.Intn(100)) * time.Millisecond) // Simulates time to consume the event.
}

func producer(event int, buffer chan int) {
	//produceEvent(event)
	startTimes[event] = time.Now()
	buffer <- event // Send event to buffer.
	//fmt.Println("Event", event, "placed on buffer.")
}

func consumer(buffer chan int, wg *sync.WaitGroup) {
	for {
		event := <-buffer // Get the event from the buffer.
		//fmt.Println("Event", event, "removed from buffer.")
		endTimes[event] = time.Now()
		//consumeEvent(event)
		wg.Done()
	}
}

func main() {
	var durations float64
	var averageWaitTimes float64
	for j := 1; j < numTrials; j++ {
		buffer := make(chan int, bufferLength) // There can only be a finite number of items on the buffer at any time.
		var wg sync.WaitGroup
		wg.Add(numProducers)
		start := time.Now()
		for i := 0; i < numConsumers; i++ {
			go consumer(buffer, &wg)
		}
		for i := 0; i < numProducers; i++ {
			go producer(i, buffer)
		}
		wg.Wait()
		end := time.Now()

		var waitTimes = endTimes[0].Sub(startTimes[0])
		for i := 1; i < numProducers; i++ {
			waitTimes += endTimes[i].Sub(startTimes[i])
		}
		averageWaitTimes += waitTimes.Seconds() / numProducers
		durations += end.Sub(start).Seconds()
	}
	fmt.Println("Average wait time:", averageWaitTimes/numTrials, "s")
	fmt.Println("Average duration:", durations/numTrials, "s")
}
